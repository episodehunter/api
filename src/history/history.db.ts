import { Db } from '../types/context.type'
import { WatchedEpisode, WatchedEpisodePage } from '../root-type'
import { watchedTableName } from '../tables'
import { safeMap } from '../util'
import { mapDatabaseWatchedEpisodeToDefinition } from '../episode/episode.util'

type GetUserEpisodeHistory = (
  userId: string,
  offset: number,
  limit: number
) => Promise<WatchedEpisode[]>

export function getGetUserEpisodeHistory(db: Db): GetUserEpisodeHistory {
  return (userId, offset, limit) => {
    return db
      .select('id', 'serie_id', 'season', 'episode', 'type', 'time')
      .from(watchedTableName)
      .where('user_id', userId)
      .orderBy('time', 'desc')
      .offset(offset)
      .limit(limit)
      .then(watched => safeMap(watched, mapDatabaseWatchedEpisodeToDefinition)) as any
  }
}

export async function getUserEpisodeHistoryPage(
  userId: string,
  offset = 0,
  limit = 100,
  getUserEpisodeHistory: GetUserEpisodeHistory
): Promise<WatchedEpisodePage> {
  const safeLimit = Math.max(Math.min(limit, 100), 0)
  const safeOffset = Math.max(Math.min(offset, 1e5), 0)
  const history = await getUserEpisodeHistory(userId, safeOffset, safeLimit + 1)
  const hasNextPage = history.length > safeLimit
  if (hasNextPage) {
    history.pop() // Remove the extra element
  }
  return {
    pageInfo: {
      hasNextPage
    },
    episodes: history
  }
}
