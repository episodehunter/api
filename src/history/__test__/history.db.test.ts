import { spy } from 'simple-spy'
import { create } from 'chain-spy'
import { getUserEpisodeHistoryPage, getGetUserEpisodeHistory } from '../history.db'

test('Set limit to max 101', async () => {
  // Arrange
  const userId = '1'
  const offset = 0
  const limit = 102
  const getUserEpisodeHistory = spy(() => Promise.resolve([]))

  // Act
  await getUserEpisodeHistoryPage(userId, offset, limit, getUserEpisodeHistory as any)

  // Assert
  expect(getUserEpisodeHistory.callCount).toBe(1)
  expect(getUserEpisodeHistory.args[0][2]).toBe(101)
})

test('Set limit to min 1', async () => {
  // Arrange
  const userId = '1'
  const offset = 0
  const limit = 0
  const getUserEpisodeHistory = spy(() => Promise.resolve([]))

  // Act
  await getUserEpisodeHistoryPage(userId, offset, limit, getUserEpisodeHistory as any)

  // Assert
  expect(getUserEpisodeHistory.callCount).toBe(1)
  expect(getUserEpisodeHistory.args[0][2]).toBe(1)
})

test('Set offset to max 100000', async () => {
  // Arrange
  const userId = '1'
  const offset = 100001
  const limit = 0
  const getUserEpisodeHistory = spy(() => Promise.resolve([]))

  // Act
  await getUserEpisodeHistoryPage(userId, offset, limit, getUserEpisodeHistory as any)

  // Assert
  expect(getUserEpisodeHistory.callCount).toBe(1)
  expect(getUserEpisodeHistory.args[0][1]).toBe(100000)
})

test('Set offset to min 0', async () => {
  // Arrange
  const userId = '1'
  const offset = -1
  const limit = 0
  const getUserEpisodeHistory = spy(() => Promise.resolve([]))

  // Act
  await getUserEpisodeHistoryPage(userId, offset, limit, getUserEpisodeHistory as any)

  // Assert
  expect(getUserEpisodeHistory.callCount).toBe(1)
  expect(getUserEpisodeHistory.args[0][1]).toBe(0)
})

test('Pass offset', async () => {
  // Arrange
  const userId = '1'
  const offset = 1
  const limit = 0
  const getUserEpisodeHistory = spy(() => Promise.resolve([]))

  // Act
  await getUserEpisodeHistoryPage(userId, offset, limit, getUserEpisodeHistory as any)

  // Assert
  expect(getUserEpisodeHistory.callCount).toBe(1)
  expect(getUserEpisodeHistory.args[0][1]).toBe(1)
})

test('Pass limit', async () => {
  // Arrange
  const userId = '1'
  const offset = 0
  const limit = 1
  const getUserEpisodeHistory = spy(() => Promise.resolve([]))

  // Act
  await getUserEpisodeHistoryPage(userId, offset, limit, getUserEpisodeHistory as any)

  // Assert
  expect(getUserEpisodeHistory.callCount).toBe(1)
  expect(getUserEpisodeHistory.args[0][2]).toBe(2)
})

test('Pass user id', async () => {
  // Arrange
  const userId = '1'
  const offset = 0
  const limit = 1
  const getUserEpisodeHistory = spy(() => Promise.resolve([]))

  // Act
  await getUserEpisodeHistoryPage(userId, offset, limit, getUserEpisodeHistory as any)

  // Assert
  expect(getUserEpisodeHistory.callCount).toBe(1)
  expect(getUserEpisodeHistory.args[0][0]).toBe('1')
})

test('Set hasNextPage to true', async () => {
  // Arrange
  const userId = '1'
  const offset = 0
  const limit = 1
  const getUserEpisodeHistory = spy(() => Promise.resolve([1, 1]))

  // Act
  const result = await getUserEpisodeHistoryPage(
    userId,
    offset,
    limit,
    getUserEpisodeHistory as any
  )

  // Assert
  expect(result.pageInfo.hasNextPage).toBe(true)
})

test('Set hasNextPage to false', async () => {
  // Arrange
  const userId = '1'
  const offset = 0
  const limit = 1
  const getUserEpisodeHistory = spy(() => Promise.resolve([1]))

  // Act
  const result = await getUserEpisodeHistoryPage(
    userId,
    offset,
    limit,
    getUserEpisodeHistory as any
  )

  // Assert
  expect(result.pageInfo.hasNextPage).toBe(false)
})

test('Pop the history when hasNextPage is true', async () => {
  // Arrange
  const userId = '1'
  const offset = 0
  const limit = 1
  const getUserEpisodeHistory = spy(() => Promise.resolve([1, 1]))

  // Act
  const result = await getUserEpisodeHistoryPage(
    userId,
    offset,
    limit,
    getUserEpisodeHistory as any
  )

  // Assert
  expect(result.episodes.length).toBe(1)
})

test('Do not pop the history when hasNextPage is false', async () => {
  // Arrange
  const userId = '1'
  const offset = 0
  const limit = 2
  const getUserEpisodeHistory = spy(() => Promise.resolve([1, 1]))

  // Act
  const result = await getUserEpisodeHistoryPage(
    userId,
    offset,
    limit,
    getUserEpisodeHistory as any
  )

  // Assert
  expect(result.episodes.length).toBe(2)
})

test('Select user episode history', () => {
  // Arrange
  const db = create()
  const userId = '1'
  const offset = 1
  const limit = 2

  // Act
  getGetUserEpisodeHistory(db)(userId, offset, limit)

  // Assert
  expect(db.__execution_log__).toMatchSnapshot()
})
