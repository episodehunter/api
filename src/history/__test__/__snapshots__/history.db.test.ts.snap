// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select user episode history 1`] = `
Array [
  Object {
    "name": "select",
    "type": "get",
  },
  Object {
    "args": Array [
      "id",
      "serie_id",
      "season",
      "episode",
      "type",
      "time",
    ],
    "type": "apply",
  },
  Object {
    "name": "from",
    "type": "get",
  },
  Object {
    "args": Array [
      "tv_watched",
    ],
    "type": "apply",
  },
  Object {
    "name": "where",
    "type": "get",
  },
  Object {
    "args": Array [
      "user_id",
      "1",
    ],
    "type": "apply",
  },
  Object {
    "name": "orderBy",
    "type": "get",
  },
  Object {
    "args": Array [
      "time",
      "desc",
    ],
    "type": "apply",
  },
  Object {
    "name": "offset",
    "type": "get",
  },
  Object {
    "args": Array [
      1,
    ],
    "type": "apply",
  },
  Object {
    "name": "limit",
    "type": "get",
  },
  Object {
    "args": Array [
      2,
    ],
    "type": "apply",
  },
  Object {
    "name": "then",
    "type": "get",
  },
  Object {
    "args": Array [
      [Function],
    ],
    "type": "apply",
  },
]
`;
